cmd/kube-controller-manager/app/core.go

```diff
@@ -688,30 +691,64 @@ func startGarbageCollectorController(ctx context.Context, controllerContext Cont
 	for _, r := range controllerContext.ComponentConfig.GarbageCollectorController.GCIgnoredResources {
 		ignoredResources[schema.GroupResource{Group: r.Group, Resource: r.Resource}] = struct{}{}
 	}
+
+	// Run and Sync dependencyGraphBuilder if not already
+	if !controllerContext.GraphBuilder.IsRunning() {
+		go controllerContext.GraphBuilder.Run(ctx)
+		go controllerContext.GraphBuilder.Sync(ctx, discoveryClient, garbagecollector.SyncPeriod)
+	}
+
+	attemptToDelete, attemptToOrphan, absentOwnerCache, eventBroadcaster := controllerContext.GraphBuilder.GetGraphResources()
 	garbageCollector, err := garbagecollector.NewGarbageCollector(
 		ctx,
 		gcClientset,
 		metadataClient,
 		controllerContext.RESTMapper,
 		ignoredResources,
 		controllerContext.ObjectOrMetadataInformerFactory,
-		controllerContext.InformersStarted,
+		controllerContext.GraphBuilder,
+		attemptToDelete,
+		attemptToOrphan,
+		absentOwnerCache,
+		eventBroadcaster,
 	)
 	if err != nil {
-		return nil, true, fmt.Errorf("failed to start the generic garbage collector: %v", err)
+		return nil, false, fmt.Errorf("failed to start the garbage collector: %w", err)
 	}
 
 	// Start the garbage collector.
 	workers := int(controllerContext.ComponentConfig.GarbageCollectorController.ConcurrentGCSyncs)
 	go garbageCollector.Run(ctx, workers)
 
-	// Periodically refresh the RESTMapper with new discovery information and sync
-	// the garbage collector.
-	go garbageCollector.Sync(ctx, discoveryClient, 30*time.Second)
-
 	return garbageCollector, true, nil
 }
 
+func setupGraphBuilder(
+	ctx context.Context,
+	metadataClient metadata.Interface,
+	informersStarted <-chan struct{},
+	restMapper *restmapper.DeferredDiscoveryRESTMapper,
+	sharedInformers informerfactory.InformerFactory,
+) *garbagecollector.GraphBuilder {
+	attemptToDelete := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_attempt_to_delete")
+	attemptToOrphan := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "garbage_collector_attempt_to_orphan")
+	absentOwnerCache := garbagecollector.NewReferenceCache(500)
+	eventBroadcaster := record.NewBroadcaster(record.WithContext(ctx))
+
```

Comment:
can't we just create these in `NewDependencyGraphBuilder` since you have the method to retrieve them later?